aggregator/src/aggregator.rs

```diff
@@ -795,7 +777,8 @@ impl<C: Clock> Aggregator<C> {
             return Err(Error::UnauthorizedRequest(*task_id));
         }
 
-        if self.clock.now() > *task_config.task_end() {
+        let task_end = task_config.task_start().add(task_config.task_duration())?;
+        if self.clock.now() > task_end {
```

Comment:
This code is implementing the requirement that "A protocol participant MUST opt out if ... [t]he task has ended."[1] I think using the server's clock, rather than the timestamp from a given report, is appropriate for this specific check.

You're correct that we should also be checking each report's timestamp against the end (and start) of the task, and we MUST reject reports that don't fall into the acceptable time interval. Those checks are implemented in `handle_upload_generic`.

Interestingly, I don't think the equivalent Helper-side timestamp check has ever been implemented, though we do implement checking that the report isn't from too far in the future. I filed https://github.com/divviup/janus/issues/3524 for this, arguing that it is not mandatory. Still, it's a small change, so I'll address it if I have time.

[1] https://www.ietf.org/archive/id/draft-ietf-ppm-dap-taskprov-01.html#name-opting-into-a-task